File: hash-table.js
class HashTable {
  constructor(size) {
    this.size = size;
    this.keys = new Array(size);
    this.values = new Array(size);
    this.limit = 0;
  }

  // Average: O(1)
  // Worst: O(n)
  put(key, value) {
    if (this.limit + 1 > this.size) throw "hash table is full";
    let hashIndex = this.hash(key);
    let squareIndex = 1;

    // Linear probing
    while (this.keys[hashIndex] != null) {
      hashIndex++;
      hashIndex %= this.size;
    }

    // Quadratic Probing
    while (this.keys[hashIndex] != null) {
      hashIndex += Math.pow(squareIndex, 2);
      hashIndex %= this.size;
      squareIndex++;
    }

    this.keys[hashIndex] = key;
    this.values[hashIndex] = value;
    this.limit++;
  }

  hash(key) {
    if (!Number.isInteger(key)) throw "must be int";
    return this.secondHash(key % this.size);
  }

  // Double Hashing
  secondHash(hashKey) {
    const remainder = this.size - 2;
    return remainder - (hashKey % remainder);
  }

  // Average: O(1)
  // Worst: O(n)
  get(key) {
    let hashIndex = this.hash(key);
    let squareIndex = 1;

    // Linear Probing
    while (this.keys[hashIndex] != key) {
      hashIndex++;
      hashIndex %= this.size;
    }

    // Quadratic Probing
    while (this.keys[hashIndex] != null) {
      hashIndex += Math.pow(squareIndex, 2);
      hashIndex %= this.size;
      squareIndex++;
    }

    return this.values[hashIndex];
  }
}

File: HashTable.py
from math import isnan

class HashTable:
  def __init__(self, size: int) -> None:
    self.size = size
    self.keys = [None] * size
    self.values = [None] * size
    self.limit = 0
  
  def put_linear_probing(self, key: int, value: int) -> None:
    if self.limit >= self.size: 
      raise Exception('Hash table is full!')
    
    hash_index = self.hash(key)

    while self.keys[hash_index]:
      hash_index += 1
      hash_index %= self.size

    self.keys[hash_index] = key
    self.values[hash_index] = value
    self.limit += 1

  def put_quadratic_probing(self, key: int, value: int) -> None:
    if self.limit >= self.size: 
      raise Exception('Hash table is full!')
    
    hash_index = self.hash(key)
    square_index = 1

    while self.keys[hash_index]:
      hash_index += pow(square_index, 2)
      square_index += 1

      hash_index %= self.size

    self.keys[hash_index] = key
    self.values[hash_index] = value
    self.limit += 1

  def get_linear_probing(self, key: int) -> int:
    hash_index = self.hash(key)
    while self.keys[hash_index] != key:
      hash_index += 1
      hash_index %= self.size
    return self.values[hash_index]

  def get_quadratic_probing(self, key: int) -> int:
    hash_index = self.hash(key)
    square_index = 1
    while self.keys[hash_index] != key:
      hash_index += pow(square_index, 2)
      square_index += 1
      hash_index %= self.size
    return self.values[hash_index]

  def second_hash(self, hashed_key: int) -> int:
    r = self.size - 2
    return r - hashed_key % r

  def hash(self, key: int) -> int:
    if isnan(key): raise Exception('Key must be an integer!')
    # Double Hashing
    return self.second_hash(key % self.size)